File: epaper_ble/android/app/src/main/java/com/epaper/ble/ble/EpaperBleClient.kt
package com.epaper.ble.ble

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.ParcelUuid
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.UUID
import java.util.concurrent.ConcurrentLinkedQueue

/** See PROTOCOL.md — one service, a control characteristic and a data characteristic. */
object Protocol {
    val SERVICE: UUID = UUID.fromString("a1b2c000-5e1f-4a7d-9c3b-2f6e8d0a4b71")
    val CONTROL: UUID = UUID.fromString("a1b2c001-5e1f-4a7d-9c3b-2f6e8d0a4b71")
    val DATA: UUID = UUID.fromString("a1b2c002-5e1f-4a7d-9c3b-2f6e8d0a4b71")
    val CCCD: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")

    const val CMD_BEGIN: Byte = 0x01
    const val CMD_COMMIT: Byte = 0x02
    const val CMD_ABORT: Byte = 0x03
    const val CMD_STATUS: Byte = 0x04

    const val EVT_STATE: Byte = 0x10
    const val EVT_PROGRESS: Byte = 0x11
    const val EVT_OK: Byte = 0x12
    const val EVT_ERROR: Byte = 0x13

    fun errorText(reason: Int): String = when (reason) {
        1 -> "No transfer open"
        2 -> "Too much data"
        3 -> "Wrong frame length"
        4 -> "Checksum mismatch"
        5 -> "Storage error on the device"
        6 -> "Display is busy"
        else -> "Unknown error $reason"
    }
}

data class DiscoveredDevice(val address: String, val name: String, val rssi: Int)

sealed interface TransferState {
    data object Idle : TransferState
    data object Scanning : TransferState
    data object Connecting : TransferState
    data object Ready : TransferState
    data class Sending(val sent: Int, val total: Int) : TransferState
    data object Refreshing : TransferState
    data object Done : TransferState
    data class Failed(val message: String) : TransferState
}

/**
 * Minimal GATT client for the frame transfer. Every GATT operation is issued
 * from the main thread and the next one waits for the previous callback, which
 * is what the Android stack expects.
 */
@SuppressLint("MissingPermission")
class EpaperBleClient(private val context: Context) {

    private val handler = Handler(Looper.getMainLooper())

    private val _devices = MutableStateFlow<List<DiscoveredDevice>>(emptyList())
    val devices: StateFlow<List<DiscoveredDevice>> = _devices.asStateFlow()

    private val _state = MutableStateFlow<TransferState>(TransferState.Idle)
    val state: StateFlow<TransferState> = _state.asStateFlow()

    private var gatt: BluetoothGatt? = null
    private var control: BluetoothGattCharacteristic? = null
    private var data: BluetoothGattCharacteristic? = null

    private val pendingChunks = ConcurrentLinkedQueue<ByteArray>()
    private var frameSize = 0
    private var sentBytes = 0
    private var chunkSize = 20

    private val adapter: BluetoothAdapter?
        get() = (context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager)?.adapter

    val isBluetoothOn: Boolean get() = adapter?.isEnabled == true

    // -------------------------------------------------------------------------
    // Scanning
    // -------------------------------------------------------------------------

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val name = result.device.name ?: result.scanRecord?.deviceName ?: return
            val found = DiscoveredDevice(result.device.address, name, result.rssi)
            _devices.value = (_devices.value.filter { it.address != found.address } + found)
                .sortedByDescending { it.rssi }
        }

        override fun onScanFailed(errorCode: Int) {
            _state.value = TransferState.Failed("Scan failed (code $errorCode)")
        }
    }

    fun startScan() {
        val scanner = adapter?.bluetoothLeScanner ?: run {
            _state.value = TransferState.Failed("Bluetooth is off")
            return
        }
        _devices.value = emptyList()
        _state.value = TransferState.Scanning

        val filter = ScanFilter.Builder().setServiceUuid(ParcelUuid(Protocol.SERVICE)).build()
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()
        scanner.startScan(listOf(filter), settings, scanCallback)

        handler.postDelayed({ stopScan() }, 15_000)
    }

    fun stopScan() {
        adapter?.bluetoothLeScanner?.stopScan(scanCallback)
        if (_state.value is TransferState.Scanning) _state.value = TransferState.Idle
    }

    // -------------------------------------------------------------------------
    // Connection
    // -------------------------------------------------------------------------

    fun connect(address: String) {
        stopScan()
        val device: BluetoothDevice = adapter?.getRemoteDevice(address) ?: run {
            _state.value = TransferState.Failed("Unknown device")
            return
        }
        _state.value = TransferState.Connecting
        gatt = device.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
    }

    fun disconnect() {
        gatt?.disconnect()
        gatt?.close()
        gatt = null
        control = null
        data = null
        pendingChunks.clear()
        _state.value = TransferState.Idle
    }

    // -------------------------------------------------------------------------
    // Transfer
    // -------------------------------------------------------------------------

    fun sendFrame(frame: ByteArray, crc: Int) {
        val controlCharacteristic = control
        if (controlCharacteristic == null || gatt == null) {
            _state.value = TransferState.Failed("Not connected")
            return
        }

        pendingChunks.clear()
        var offset = 0
        while (offset < frame.size) {
            val end = minOf(offset + chunkSize, frame.size)
            pendingChunks.add(frame.copyOfRange(offset, end))
            offset = end
        }

        frameSize = frame.size
        sentBytes = 0
        _state.value = TransferState.Sending(0, frameSize)

        val begin = ByteArray(9)
        begin[0] = Protocol.CMD_BEGIN
        writeIntLe(begin, 1, frame.size)
        writeIntLe(begin, 5, crc)
        writeCharacteristic(controlCharacteristic, begin, withResponse = true)
    }

    private fun sendNextChunk() {
        val chunk = pendingChunks.poll()
        val dataCharacteristic = data
        if (chunk == null || dataCharacteristic == null) {
            val controlCharacteristic = control ?: return
            writeCharacteristic(
                controlCharacteristic,
                byteArrayOf(Protocol.CMD_COMMIT),
                withResponse = true
            )
            return
        }
        writeCharacteristic(dataCharacteristic, chunk, withResponse = true)
    }

    private fun writeCharacteristic(
        characteristic: BluetoothGattCharacteristic,
        value: ByteArray,
        withResponse: Boolean
    ) {
        val currentGatt = gatt ?: return
        val writeType = if (withResponse) {
            BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
        } else {
            BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE
        }

        handler.post {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                currentGatt.writeCharacteristic(characteristic, value, writeType)
            } else {
                @Suppress("DEPRECATION")
                characteristic.writeType = writeType
                @Suppress("DEPRECATION")
                characteristic.value = value
                @Suppress("DEPRECATION")
                currentGatt.writeCharacteristic(characteristic)
            }
        }
    }

    private fun writeIntLe(target: ByteArray, offset: Int, value: Int) {
        target[offset] = (value and 0xFF).toByte()
        target[offset + 1] = (value shr 8 and 0xFF).toByte()
        target[offset + 2] = (value shr 16 and 0xFF).toByte()
        target[offset + 3] = (value shr 24 and 0xFF).toByte()
    }

    // -------------------------------------------------------------------------
    // GATT callbacks
    // -------------------------------------------------------------------------

    private val gattCallback = object : BluetoothGattCallback() {

        override fun onConnectionStateChange(g: BluetoothGatt, status: Int, newState: Int) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                g.requestMtu(247)
            } else {
                val wasTransferring = _state.value is TransferState.Sending
                handler.post {
                    if (wasTransferring) {
                        _state.value = TransferState.Failed("Disconnected during transfer")
                    } else if (_state.value !is TransferState.Done) {
                        _state.value = TransferState.Idle
                    }
                }
                g.close()
                gatt = null
            }
        }

        override fun onMtuChanged(g: BluetoothGatt, mtu: Int, status: Int) {
            // Three bytes of ATT header come off the top of every write.
            chunkSize = (mtu - 3).coerceIn(20, 512)
            g.discoverServices()
        }

        override fun onServicesDiscovered(g: BluetoothGatt, status: Int) {
            val service = g.getService(Protocol.SERVICE)
            if (service == null) {
                handler.post { _state.value = TransferState.Failed("Service not found") }
                return
            }
            control = service.getCharacteristic(Protocol.CONTROL)
            data = service.getCharacteristic(Protocol.DATA)

            val controlCharacteristic = control
            if (controlCharacteristic == null || data == null) {
                handler.post { _state.value = TransferState.Failed("Characteristics not found") }
                return
            }

            g.setCharacteristicNotification(controlCharacteristic, true)
            val cccd = controlCharacteristic.getDescriptor(Protocol.CCCD)
            if (cccd != null) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    g.writeDescriptor(cccd, BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE)
                } else {
                    @Suppress("DEPRECATION")
                    cccd.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                    @Suppress("DEPRECATION")
                    g.writeDescriptor(cccd)
                }
            } else {
                handler.post { _state.value = TransferState.Ready }
            }
        }

        override fun onDescriptorWrite(g: BluetoothGatt, d: BluetoothGattDescriptor, status: Int) {
            handler.post { _state.value = TransferState.Ready }
        }

        override fun onCharacteristicWrite(
            g: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            status: Int
        ) {
            if (status != BluetoothGatt.GATT_SUCCESS) {
                handler.post { _state.value = TransferState.Failed("Write failed (status $status)") }
                return
            }

            when (characteristic.uuid) {
                Protocol.DATA -> {
                    sentBytes = minOf(sentBytes + chunkSize, frameSize)
                    handler.post { _state.value = TransferState.Sending(sentBytes, frameSize) }
                    sendNextChunk()
                }

                Protocol.CONTROL -> {
                    // BEGIN was acknowledged; COMMIT is answered by a notification.
                    if (pendingChunks.isNotEmpty()) sendNextChunk()
                }
            }
        }

        @Deprecated("Kept for API < 33")
        @Suppress("DEPRECATION")
        override fun onCharacteristicChanged(
            g: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic
        ) {
            handleNotification(characteristic.value ?: return)
        }

        override fun onCharacteristicChanged(
            g: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            value: ByteArray
        ) {
            handleNotification(value)
        }
    }

    private fun handleNotification(value: ByteArray) {
        if (value.isEmpty()) return
        when (value[0]) {
            Protocol.EVT_OK -> handler.post { _state.value = TransferState.Done }

            Protocol.EVT_ERROR -> {
                val reason = if (value.size > 1) value[1].toInt() else 0
                handler.post { _state.value = TransferState.Failed(Protocol.errorText(reason)) }
            }

            Protocol.EVT_STATE -> {
                if (value.size > 1 && value[1].toInt() == 3) {
                    handler.post { _state.value = TransferState.Refreshing }
                }
            }
        }
    }
}

File: epaper_ble/ios/EpaperBLE/EpaperBleClient.swift
import CoreBluetooth
import Foundation

/// See PROTOCOL.md — one service, a control characteristic and a data one.
enum Protocolo {
    static let service = CBUUID(string: "a1b2c000-5e1f-4a7d-9c3b-2f6e8d0a4b71")
    static let control = CBUUID(string: "a1b2c001-5e1f-4a7d-9c3b-2f6e8d0a4b71")
    static let data = CBUUID(string: "a1b2c002-5e1f-4a7d-9c3b-2f6e8d0a4b71")

    static let cmdBegin: UInt8 = 0x01
    static let cmdCommit: UInt8 = 0x02
    static let cmdAbort: UInt8 = 0x03

    static let evtState: UInt8 = 0x10
    static let evtProgress: UInt8 = 0x11
    static let evtOk: UInt8 = 0x12
    static let evtError: UInt8 = 0x13

    static func errorText(_ reason: Int) -> String {
        switch reason {
        case 1: return "No transfer open"
        case 2: return "Too much data"
        case 3: return "Wrong frame length"
        case 4: return "Checksum mismatch"
        case 5: return "Storage error on the device"
        case 6: return "Display is busy"
        default: return "Unknown error \(reason)"
        }
    }
}

struct DiscoveredDevice: Identifiable, Equatable {
    let id: UUID
    let name: String
    let rssi: Int
}

enum TransferState: Equatable {
    case idle
    case scanning
    case connecting
    case ready
    case sending(sent: Int, total: Int)
    case refreshing
    case done
    case failed(String)

    var text: String {
        switch self {
        case .idle: return "Not connected"
        case .scanning: return "Scanning…"
        case .connecting: return "Connecting…"
        case .ready: return "Connected"
        case .sending(let sent, let total): return "Sending \(sent) / \(total) bytes"
        case .refreshing: return "Refreshing the panel…"
        case .done: return "Shown on the display"
        case .failed(let message): return "Error: \(message)"
        }
    }

    var isConnected: Bool {
        switch self {
        case .ready, .sending, .refreshing, .done: return true
        default: return false
        }
    }
}

@MainActor
final class EpaperBleClient: NSObject, ObservableObject {

    @Published private(set) var devices: [DiscoveredDevice] = []
    @Published private(set) var state: TransferState = .idle

    private var manager: CBCentralManager!
    private var peripheral: CBPeripheral?
    private var control: CBCharacteristic?
    private var data: CBCharacteristic?
    private var discovered: [UUID: CBPeripheral] = [:]

    private var chunks: [Data] = []
    private var chunkIndex = 0
    private var sentBytes = 0
    private var frameSize = 0

    override init() {
        super.init()
        manager = CBCentralManager(delegate: self, queue: .main)
    }

    func startScan() {
        guard manager.state == .poweredOn else {
            state = .failed("Bluetooth is off")
            return
        }
        devices = []
        discovered = [:]
        state = .scanning
        manager.scanForPeripherals(withServices: [Protocolo.service])

        // Scanning forever is a battery drain; the user can scan again.
        DispatchQueue.main.asyncAfter(deadline: .now() + 15) { [weak self] in
            self?.stopScan()
        }
    }

    func stopScan() {
        manager.stopScan()
        if state == .scanning { state = .idle }
    }

    func connect(_ id: UUID) {
        guard let target = discovered[id] else {
            state = .failed("Unknown device")
            return
        }
        stopScan()
        state = .connecting
        peripheral = target
        target.delegate = self
        manager.connect(target)
    }

    func disconnect() {
        if let peripheral { manager.cancelPeripheralConnection(peripheral) }
        peripheral = nil
        control = nil
        data = nil
        chunks = []
        state = .idle
    }

    func send(frame: Data) {
        guard let peripheral, let control else {
            state = .failed("Not connected")
            return
        }

        // Three bytes of ATT header come off the top of every write.
        let chunkSize = min(peripheral.maximumWriteValueLength(for: .withResponse), 244)
        chunks = stride(from: 0, to: frame.count, by: chunkSize).map {
            frame.subdata(in: $0..<min($0 + chunkSize, frame.count))
        }
        chunkIndex = 0
        sentBytes = 0
        frameSize = frame.count
        state = .sending(sent: 0, total: frameSize)

        let crc = Dither.crc32(frame)
        var payload = Data([Protocolo.cmdBegin])
        payload.append(contentsOf: withUnsafeBytes(of: UInt32(frame.count).littleEndian) { Array($0) })
        payload.append(contentsOf: withUnsafeBytes(of: crc.littleEndian) { Array($0) })
        peripheral.writeValue(payload, for: control, type: .withResponse)
    }

    private func sendNextChunk() {
        guard let peripheral, let data, let control else { return }
        if chunkIndex < chunks.count {
            peripheral.writeValue(chunks[chunkIndex], for: data, type: .withResponse)
        } else {
            peripheral.writeValue(Data([Protocolo.cmdCommit]), for: control, type: .withResponse)
        }
    }
}

extension EpaperBleClient: CBCentralManagerDelegate {

    nonisolated func centralManagerDidUpdateState(_ central: CBCentralManager) {
        Task { @MainActor in
            if central.state != .poweredOn, self.state == .scanning {
                self.state = .failed("Bluetooth is off")
            }
        }
    }

    nonisolated func centralManager(
        _ central: CBCentralManager,
        didDiscover peripheral: CBPeripheral,
        advertisementData: [String: Any],
        rssi RSSI: NSNumber
    ) {
        let name = peripheral.name
            ?? advertisementData[CBAdvertisementDataLocalNameKey] as? String
            ?? "Unknown"
        let found = DiscoveredDevice(id: peripheral.identifier, name: name, rssi: RSSI.intValue)
        Task { @MainActor in
            self.discovered[peripheral.identifier] = peripheral
            self.devices = (self.devices.filter { $0.id != found.id } + [found])
                .sorted { $0.rssi > $1.rssi }
        }
    }

    nonisolated func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        peripheral.discoverServices([Protocolo.service])
    }

    nonisolated func centralManager(
        _ central: CBCentralManager,
        didFailToConnect peripheral: CBPeripheral,
        error: Error?
    ) {
        Task { @MainActor in self.state = .failed("Could not connect") }
    }

    nonisolated func centralManager(
        _ central: CBCentralManager,
        didDisconnectPeripheral peripheral: CBPeripheral,
        error: Error?
    ) {
        Task { @MainActor in
            if case .sending = self.state {
                self.state = .failed("Disconnected during transfer")
            } else if self.state != .done {
                self.state = .idle
            }
        }
    }
}

extension EpaperBleClient: CBPeripheralDelegate {

    nonisolated func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard let service = peripheral.services?.first(where: { $0.uuid == Protocolo.service }) else {
            Task { @MainActor in self.state = .failed("Service not found") }
            return
        }
        peripheral.discoverCharacteristics([Protocolo.control, Protocolo.data], for: service)
    }

    nonisolated func peripheral(
        _ peripheral: CBPeripheral,
        didDiscoverCharacteristicsFor service: CBService,
        error: Error?
    ) {
        let found = service.characteristics ?? []
        let controlCharacteristic = found.first { $0.uuid == Protocolo.control }
        let dataCharacteristic = found.first { $0.uuid == Protocolo.data }

        guard let controlCharacteristic, let dataCharacteristic else {
            Task { @MainActor in self.state = .failed("Characteristics not found") }
            return
        }

        Task { @MainActor in
            self.control = controlCharacteristic
            self.data = dataCharacteristic
        }
        // The firmware relies on the CCCD NimBLE creates automatically.
        peripheral.setNotifyValue(true, for: controlCharacteristic)
    }

    nonisolated func peripheral(
        _ peripheral: CBPeripheral,
        didUpdateNotificationStateFor characteristic: CBCharacteristic,
        error: Error?
    ) {
        Task { @MainActor in
            if let error {
                self.state = .failed("Could not enable notifications: \(error.localizedDescription)")
            } else {
                self.state = .ready
            }
        }
    }

    nonisolated func peripheral(
        _ peripheral: CBPeripheral,
        didWriteValueFor characteristic: CBCharacteristic,
        error: Error?
    ) {
        if let error {
            Task { @MainActor in self.state = .failed("Write failed: \(error.localizedDescription)") }
            return
        }
        let isData = characteristic.uuid == Protocolo.data
        Task { @MainActor in
            if isData {
                self.sentBytes = min(self.sentBytes + (self.chunks.first?.count ?? 0), self.frameSize)
                self.chunkIndex += 1
                self.state = .sending(sent: self.sentBytes, total: self.frameSize)
                self.sendNextChunk()
            } else if self.chunkIndex < self.chunks.count {
                // BEGIN was acknowledged; COMMIT is answered by a notification.
                self.sendNextChunk()
            }
        }
    }

    nonisolated func peripheral(
        _ peripheral: CBPeripheral,
        didUpdateValueFor characteristic: CBCharacteristic,
        error: Error?
    ) {
        guard let value = characteristic.value, let opcode = value.first else { return }
        let bytes = [UInt8](value)
        Task { @MainActor in
            switch opcode {
            case Protocolo.evtOk:
                self.state = .done
            case Protocolo.evtError:
                self.state = .failed(Protocolo.errorText(bytes.count > 1 ? Int(bytes[1]) : 0))
            case Protocolo.evtState:
                if bytes.count > 1, bytes[1] == 3 { self.state = .refreshing }
            default:
                break
            }
        }
    }
}
